### PR TITLE
Ensure that setup_class failures are recorded before trying to execut…

### PR DESCRIPTION
**Pull request by dthkao**
Ensure that setup_class failures are recorded before trying to execut…e teardown_class

---
This change is [Reviewable](https://reviewable.io/reviews/google/mobly/236)

### CODE REVIEW

**Comment by adorokhine** [comment]
Review status: 0 of 1 files reviewed at latest revision, 1 unresolved discussion.

---

*[mobly/base_test.py, line 601 at r1](https://reviewable.io:443/reviews/google/mobly/236#-Kn6YDO7aiAv8bNYrzsa:-Kn6YDO81iVTyaPKeNAX:b-ijek44) ([raw file](https://github.com/google/mobly/blob/366356b0da0e6719c8b4111a5776cf1495c88dcb/mobly/base_test.py#L601)):*
> ```Python
>             self._exec_procedure_func(self._on_fail, class_record)
>             self.results.fail_class(class_record)
>             self._safe_exec_func(self.teardown_class)
> ```

Would it be possible to add a unit test for this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/236)*

**Comment by dthkao** [comment]
Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[mobly/base_test.py, line 601 at r1](https://reviewable.io:443/reviews/google/mobly/236#-Kn6YDO7aiAv8bNYrzsa:-Kn6qdxIWnOYC59liJl7:b-896fix) ([raw file](https://github.com/google/mobly/blob/366356b0da0e6719c8b4111a5776cf1495c88dcb/mobly/base_test.py#L601)):*
<details><summary><i>Previously, adorokhine (Alexander Dorokhine) wrote…</i></summary><blockquote>

Would it be possible to add a unit test for this?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/236)*

**Comment by adorokhine** [comment]
Looks like travis CI fails

---

Reviewed 1 of 1 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/236#-:-Kn6td8ft0sGfZSq4ADD:bh1thl2)*

**Comment by dthkao** [comment]
PTAL

---

Review status: 1 of 2 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/236#-:-Kn7737xPUol80KnixbB:bb71pro)*

**Comment by adorokhine** [comment]
:lgtm:

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/google/mobly/236#-:-Kn77hDWrYrmp6vACbS3:bnfp4nl)*

File: mobly/base_test.py [hunk 1]
```diff
@@ -597,8 +597,8 @@ def run(self, test_names=None):
             class_record.test_begin()
             class_record.test_fail(e)
             self._exec_procedure_func(self._on_fail, class_record)
-            self._safe_exec_func(self.teardown_class)
             self.results.fail_class(class_record)
+            self._safe_exec_func(self.teardown_class)
```
Comment (xpconanfan) [inline mobly/base_test.py]: Curious, if teardown class fails, does it get recorded?...

Comment (dthkao) [inline mobly/base_test.py]: teardown_class wasn't recorded before this change.

Comment (xpconanfan) [inline mobly/base_test.py]: Yeah, I know.
Just wondering if this has been a bug all along...

**Pull request merged**